pkg/sql/copy.go

```diff
@@ -383,6 +387,15 @@ func (c *copyMachine) readTextData(ctx context.Context, final bool) (brk bool, e
 }
 
 func (c *copyMachine) readCSVData(ctx context.Context, final bool) (brk bool, err error) {
+	// Don't try to parse the CSV until we read all the data. This is because
+	// newlines can appear in the middle of CSV input if it's quoted, so
+	// we don't have a good way of knowing how many bytes to read for the next
+	// record ahead of time (without reimplementing all the CSV parsing logic
+	// here). This means the input buffer will keep growing -- the memory is
+	// accounted in processCopyData.
```

Comment:
So what will happen if I want to upload a CSV file that's 10GB? It won't process a single row from it until all 10GBs are uploaded to memory? If so, that just trades one problem for another.

Reply:
I agree. I was thinking about how to do this and I think we'll need to fork the csv reader implementation from go. (we'd also need to do that anyway in order to support some of the other COPY options)

one added challenge is that the COPY CSV format requires a newline to appear after each record *except* for the last one. so if we encounter an EOF on an individual CopyData input buffer, it's hard to tell if that's the end of all the CSV input or if we need to check for another input buffer

tomorrow I'll take a look at how postgres does this and explore some other CSV parsers in go.

Reply:
Would it help if I create a separate ticket just for partial records with COPY CSV with the info in my [comment](https://github.com/cockroachdb/cockroach/issues/68805#issuecomment-897891364)? That way this ticket could be merged sooner while the bigger problem with the CSV reader could be worked on separately?

Reply:
yes, sure another ticket would help. if i don't get to this soon then we can go ahead with this PR without the CSV buffering changes

I did find that Postgres solves this by buffering line-by-line, and taking into account whether it is in a quoted field or not: https://github.com/postgres/postgres/blob/9626325da5e8e23ff90091bc96535495d350f06e/src/backend/commands/copyfromparse.c#L1090-L1096